### DISCUSSION TITLE
Defining the request body type in a dynamically created endpoint - typing issues

### DISCUSSION THREAD

**Discussion by provinzkraut** [fastapi/fastapi#6893] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from typing import Type
from pydantic import BaseModel
from fastapi import FastAPI

app = FastAPI()


def create_endpoint(model: Type[BaseModel]):
    @app.post("/")
    def post(data: model):
        pass
```

### Description

* Run mypy against this snippet
* Receive the error `Variable "model" is not valid as a type`

The provided example works as intended, but mypy - rightfully - complains about this since type variables cannot be used as types. The problem however is that using `Type[...]` to annotate `create_endpoint` is correct but there's no way (that I am aware of) in FastAPI to define the type of the request body other than with an annotation.

It would be really great to have a different way to achieve this, because as of now, dynamically creating endpoints like this is a bit inconvenient. 

### Environment

* OS: Debian 10.8
* FastAPI Version: 0.63.0
* Python version: 3.8.6

**Comment by Kludex** [comment]
There's a difference between runtime and when the type checking happens.

References:
- [How do I type a generic factory function?](https://github.com/python/mypy/issues/6073)
- [Useful variable to understand `type checking`](https://mypy.readthedocs.io/en/latest/runtime_troubles.html#typing-type-checking)

Maybe using [Generic class](https://mypy.readthedocs.io/en/stable/generics.html) can help you?

**Comment by provinzkraut** [comment]
@Kludex 
Thank you for your suggestions!

I am aware of the differences between runtime typed and type checking, I'd just like to get them to play along nicely. Because as of now, I can either satisfy the type checker (i.e. using generics) *or* define the endpoint in a way FastAPI understands.

Could you elaborate how you think using generics could solve this problem?

Because so far, every approach I've tried leads to FastAPI not receiving the correct type for the body parameter.

**Comment by Kludex** [comment]
I got some issues with the generics approach... And then I recalled that I've done something "similar".

Would you be satisfied with this https://github.com/Kludex/fastapi-health/blob/main/fastapi_health/route.py ? 

You'd need to change the implementation, but what I wanted to show you is the idea of it.

EDIT: I'm going to copy paste here because I don't know if I'll change that package later on:
```python
from inspect import Parameter, Signature
from typing import Callable, List

from fastapi import Depends, Response


def health(conditions: List[Callable]):
    def endpoint(**dependencies):
        if all([dependency for dependency in dependencies.values()]):
            return Response(status_code=200)
        return Response(status_code=503)

    params = []
    for condition in conditions:
        params.append(
            Parameter(
                f"param_{condition.__name__}",
                kind=Parameter.POSITIONAL_OR_KEYWORD,
                annotation=bool,
                default=Depends(condition),
            )
        )
    endpoint.__signature__ = Signature(params)
    return endpoint
```

**Comment by provinzkraut** [comment]
That's a very nice solution and indeed an approach that I can use. I'd just hoped that I wouldn't have to resort to such measures to achieve such a simple task. 

It'd be great if FastAPI could provide a solution for this. Otherwise, maybe this could be included in the docs somewhere because I do think this is a relatively common thing to do.

**Comment by andrew222651** [comment]
I don't know if this should be closed since the given solution doesn't have checkable type annotations. You might as well stick with the original code and tell mypy to ignore the line.

See also https://github.com/microsoft/pyright/discussions/3510

**Comment by JacobHayes** [comment]
I think this is a pretty clean approach with only a bit of hackery:
```python
from typing import TYPE_CHECKING, Annotated, Any

from fastapi import FastAPI
from pydantic import BaseModel

if TYPE_CHECKING:
    Concrete = Annotated
else:
    type Static = Any
    type Runtime = Any

    class Concrete:
        def __class_getitem__(cls, args: tuple[Static, Runtime]):
            return args[1]  # TODO: If static is a bound type var, check that runtime matches.


# End of helper code #

app = FastAPI()


def create_endpoint[T: BaseModel](model: type[T]):
    @app.post(f"/{model.__name__.lower()}")
    def post(data: Concrete[T, model]) -> Concrete[T, model]:
        return data

    return post
```

See [below](https://github.com/fastapi/fastapi/discussions/6893#discussioncomment-12333870) for a full example.

---

<details>
<summary>Old approach</summary>

I put together another version that works by using normal TypeVars that type checkers expect in the actual function signature, but then decorating it with a helper function that generates a `__signature__` with the actual runtime types that FastAPI expects:

```python
from collections.abc import Callable
from inspect import Parameter, Signature
from typing import Any, reveal_type

import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel


def _build_param(name: str, value: type[Any] | tuple[type[Any], object]):
    annotation, default = value, Parameter.empty
    if isinstance(value, tuple):
        annotation, default = value
    return Parameter(name, kind=Parameter.KEYWORD_ONLY, annotation=annotation, default=default)


# NOTE: This could be typed more accurately with the pending PEP 747 TypeForm.
def sig(return_annotation: type[Any] | None = None, **kwargs: type[Any] | tuple[type[Any], object]):
    def decorate[**P, R](fn: Callable[P, R]) -> Callable[P, R]:
        fn.__signature__ = Signature(  # pyright: ignore[reportFunctionMemberAccess]
            [_build_param(name, value) for name, value in kwargs.items()],
            return_annotation=return_annotation,
        )
        return fn

    return decorate

# End of helper code #

app = FastAPI()


def create_endpoint[T: BaseModel](model: type[T]):
    @app.post(f"/{model.__name__.lower()}")
    @sig(data=model, return_annotation=model)
    def post(data: T):
        return data

    return post


class Foo(BaseModel):
    x: int


class Bar(BaseModel):
    x: str


foo = create_endpoint(Foo)
reveal_type(foo)  # (data: Foo) -> Foo
foo(Foo(x=5))
foo(Bar(x="hi"))  # type error: "Bar" is not assignable to "Foo"

bar = create_endpoint(Bar)
reveal_type(bar)  # (data: Bar) -> Bar
bar(Bar(x="hi"))
bar(Foo(x=5))  # type error: "Foo" is not assignable to "Bar"

if __name__ == "__main__":
    uvicorn.run(app)
```

This makes both the type checker and FastAPI happy. There is no actual static nor runtime validation that types passed to `@sig` match the function (though you could add some runtime checks with more work), but at the least the two definitions sit next to each other. It could also probably be extended to pull existing annotations from the actual function (eg: for non-variable hints) so you don't have to pass them _all_ to `@sig`.

If you want dependencies, you can pass in `@sig(param=Annotated[..., Depends()])`. If you want default values, you can pass in a tuple like `@sig(param_with_default=(int, 5))`.

</details>

**Reply by JacobHayes** [reply]
Ok, I think this can be much simpler (and quite seamless) with some `__class_getitem__` hackery and `if TYPE_CHECKING`/`Annotated` trickery:

```python
from typing import TYPE_CHECKING, Annotated, Any, reveal_type

import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel

if TYPE_CHECKING:
    Concrete = Annotated
else:
    type Static = Any
    type Runtime = Any

    class Concrete:
        def __class_getitem__(cls, args: tuple[Static, Runtime]):
            return args[1]  # TODO: If static is a bound type var, check that runtime matches.

# End of helper code #

app = FastAPI()


def create_endpoint[T: BaseModel](model: type[T]):
    @app.post(f"/{model.__name__.lower()}")
    def post(data: Concrete[T, model]) -> Concrete[T, model]:
        return data

    return post


class Foo(BaseModel):
    x: int


foo = Foo(x=5)


class Bar(BaseModel):
    x: str


bar = Bar(x="hi")

f = create_endpoint(Foo)
f(foo)
f(bar)  # type error: "Bar" is not assignable to "Foo"
reveal_type(f)  # (data: Foo) -> Foo

b = create_endpoint(Bar)
b(bar)
b(foo)  # type error: "Foo" is not assignable to "Bar"
reveal_type(b)  # (data: Bar) -> Bar

if __name__ == "__main__":
    uvicorn.run(app)
```

With this, `Concrete[T, var]` is:
- considered by type checkers to be `T` (a valid type hint, not a variable)
- but resolved to exactly `var` at runtime (the actual type FastAPI wants)
